SpringProject/src/main/java/com/gradproject2019/users/service/UserServiceImpl.java

```diff
@@ -118,7 +115,16 @@ private void checkIfManagerExists(Long managerId) {
 
     private void checkTokenMatchesUser(UUID token, Long userId) {
         if (!authServiceImpl.getTokenById(token).getUserId().equals(userId)) {
-            throw new UserUnauthorisedException();
+            throw new UserForbiddenException();
+        }
+    }
+
+    private void checkUserRequestingIsAuthorized(Long requestedUserId, UUID token) {
+        User requestingUser = getUserById(authServiceImpl.getTokenById(token).getUserId());
+        User requestedUser = getUserById(requestedUserId);
```

Comment:
I'd suggest doing this in MySQL - you can merge this because I assume it works how you'd expect, but sequential queries like this scale poorly. e.g. 

```

SELECT EXISTS (
    SELECT 1
    FROM users
    WHERE manager_id = :requestingUserId
    OR id = :requestingUserManagerId
);
```
which will return as a Boolean that you can throw you're exception based on this

I don't understand `checkTokenMatchesUser(token, requestedUserId);` it looks like you are checking the user whose details are being requested has logged in to your app - presumably these tokens expire so what happens if this user hasn't logged in in a while? This might be me failing to understand your architecture ... 

Btw to get your user from an auth token on 1 trip to the db (you want to do this) try 

```
SELECT (your,cols,etc)
FROM users usrs
INNER JOIN tokens tkns ON (usrs.id = tkns.user_id)
WHERE tkns.token = :tokenIncoming
```

You might have heard some stuff about JOINS being expensive - in this case joining 2 columns with autoincrementing INTs will be very quick - think about how you might find an element in an ordered list/array vs unordered in Java to understand why this might be quick 

(you don't need to explictly write the `INNER` but I'm including it to make this comment make more sense)